yt/utilities/lib/pixelization_routines.pyx

```diff
@@ -1154,6 +1154,136 @@ def pixelize_sph_kernel_slice(
                     continue
                 buff[xi, yi] += buff_num[xi, yi] / buff_denom[xi, yi]
 
+@cython.initializedcheck(False)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def pixelize_sph_kernel_arbitrary_grid(np.float64_t[:, :, :] buff,
+        np.float64_t[:] posx, np.float64_t[:] posy, np.float64_t[:] posz,
+        np.float64_t[:] hsml, np.float64_t[:] pmass,
+        np.float64_t[:] pdens,
+        np.float64_t[:] quantity_to_smooth,
+        bounds, kernel_name="cubic",
+        use_normalization=True):
+
+    cdef np.intp_t xsize, ysize, zsize
+    cdef np.float64_t x_min, x_max, y_min, y_max, z_min, z_max, w_j, coeff
+    cdef np.int64_t xi, yi, zi, x0, x1, y0, y1, z0, z1
+    cdef np.float64_t qxy, posx_diff, posy_diff, posz_diff
+    cdef np.float64_t x, y, z, dx, dy, dz, idx, idy, idz, h_j3, h_j2, h_j, ih_j2
+    cdef int index, i, j, k
+    cdef np.float64_t[:, :, :] buff_num
+    cdef np.float64_t[:, :, :] buff_denom
+
+    xsize, ysize, zsize = buff.shape[0], buff.shape[1], buff.shape[2]
+    buff_num = np.zeros((xsize, ysize, zsize), dtype="float64")
+    buff_denom = np.zeros((xsize, ysize, zsize), dtype="float64")
+
+    x_min = bounds[0]
+    x_max = bounds[1]
+    y_min = bounds[2]
+    y_max = bounds[3]
+    z_min = bounds[4]
+    z_max = bounds[5]
+
+    dx = (x_max - x_min) / xsize
+    dy = (y_max - y_min) / ysize
+    dz = (z_max - z_min) / zsize
+    idx = 1.0/dx
+    idy = 1.0/dy
+    idz = 1.0/dz
+
+    kernel_func = get_kernel_func(kernel_name)
+
+    # define this to avoid using the use_normalization python object
+    # in the tight loop
+    cdef np.intp_t use_norm = int(use_normalization)
+
+    pbar = get_pbar("Interpolating particle properties onto 3D grid",
+        posx.shape[0],parallel=True)
+    with nogil:
+        for j in range(0, posx.shape[0]):
+            if j % 1000 == 0:
```

Comment:
in response to your last comment, have you tried a smaller number here? Also keep in mind that most people will be generating covering grids that are at least ~32^3, often much bigger (bigger than the grids you've been testing with, i think)

Reply:
yeah! I should change that! I should likely change it in the projection function too

Reply:
Just be careful to measure whether you’re introducing noticeable overhead (e.g. by timing a benchmark while varying this).